src/nc/core/likec/Simplifier.cpp

```diff
@@ -400,7 +419,48 @@ std::unique_ptr<MemberAccessOperator> Simplifier::simplify(std::unique_ptr<Membe
 
 std::unique_ptr<Expression> Simplifier::simplify(std::unique_ptr<Typecast> node) {
     node->operand() = simplify(std::move(node->operand()));
-
+    /*
+     * simplify casts within casts
+     * woooooo! this works!
+     * and, it has a fairly noticeable impact
+     */
+    if(node->operand()->is<UnaryOperator>()
+                                          && node->type()->isInteger() &&
+            node->operand()->as<UnaryOperator>()->operatorKind() == UnaryOperator::DEREFERENCE) {
+        /*
+         * node's operand is dereference. node is a cast to an IntegerType
+         */
+        auto deref = node->operand()->as<UnaryOperator>();
+
+        if(deref->operand()->is<Typecast>()) {
+            /*
+             * we now know that this cast's operand is a dereference
+             * of a casted pointer
+             *
+             * we're going to check to see if the type of pointer it's casted to
+             * is compatible with the type this node is casting to
+             * i.e reinterpret_cast<int16_t>(*reinterpret_cast<uint16_t*>(&edi2))
+             * will become *reinterpret_cast<int16_t*>(&edi2)
+             */
+
+            auto innerCast = deref->operand()->as<Typecast>();
+            if (innerCast->type()->isPointer() && !node->type()->isPointer()) {
+                //inner cast is a pointer type
+                auto ptrType = innerCast->type()->as<PointerType>();
+                if (ptrType->pointeeType()->size() == node->type()->size()
+                    && ptrType->pointeeType()->isInteger()) {
+                    
+                    auto integerPointee = ptrType->pointeeType()->as<IntegerType>();
+
+                    deref->operand() =
+                    std::make_unique<Typecast>(Typecast::CastKind::REINTERPRET_CAST, new PointerType(
```

Comment:
Do not create types yourself, ask likec::Tree to do it for you.
It will own the types it returns; it will also avoid creating the same type twice.
The code as written has a memory leak.